variable/include/scipp/variable/variable.tcc

```diff
@@ -271,411 +51,131 @@ struct default_init<Eigen::Matrix<T, Rows, Cols>> {
-    } else {
-      return {*m_variances, dims};
-    }
-  }
-
-  std::unique_ptr<
-      VariableConceptT<std::remove_const_t<typename T::element_type>>>
-  copyT() const override {
-    using DataT = element_array<std::remove_const_t<typename T::element_type>>;
-    DataT values(m_values.begin(), m_values.end());
-    std::optional<DataT> variances;
-    if (hasVariances())
-      variances = DataT(m_variances->begin(), m_variances->end());
-    return std::make_unique<DataModel<DataT>>(this->dims(), std::move(values),
-                                              std::move(variances));
-  }
-
-  VariableConceptHandle clone() const override {
-    return std::make_unique<ViewModel<T>>(this->dims(), m_values, m_variances);
-  }
+VariableConceptHandle
+DataModel<T>::makeDefaultFromParent(const Dimensions &dims) const {
+  if (hasVariances())
+    return std::make_unique<DataModel<T>>(dims, element_array<T>(dims.volume()),
+                                          element_array<T>(dims.volume()));
+  else
+    return std::make_unique<DataModel<T>>(dims,
+                                          element_array<T>(dims.volume()));
+}
 
-  bool isSame(const VariableConcept &other) const override {
-    // Views can be copied but can still refer to same data, so unlike for
-    // DataModel a pointer comparison is not sufficient here.
-    if (hasVariances() != other.hasVariances())
-      return false;
-    if (const auto *ptr = dynamic_cast<const ViewModel<T> *>(&other))
-      return m_values.isSame(ptr->m_values);
+/// Helper for implementing Variable(View)::operator==.
+///
+/// This method is using virtual dispatch as a trick to obtain T, such that
+/// values<T> and variances<T> can be compared.
+template <class T>
+bool DataModel<T>::equals(const VariableConstView &a,
+                          const VariableConstView &b) const {
+  if (a.unit() != b.unit())
     return false;
-  }
-
-  bool isContiguous() const override {
-    return this->dims().isContiguousIn(m_values.parentDimensions());
-  }
-  bool isView() const override { return true; }
-  bool isConstView() const override {
-    return std::is_const<typename T::element_type>::value;
-  }
-  bool hasVariances() const noexcept override {
-    return m_variances.has_value();
-  }
+  if (a.dims() != b.dims())
+    return false;
+  if (a.dtype() != b.dtype())
+    return false;
+  if (a.hasVariances() != b.hasVariances())
+    return false;
+  if (a.dims().volume() == 0 && a.dims() == b.dims())
+    return true;
+  return equal(a.values<T>(), b.values<T>()) &&
+         (!a.hasVariances() || equal(a.variances<T>(), b.variances<T>()));
+}
 
-  scipp::index size() const override { return m_values.size(); }
+/// Helper for implementing Variable(View) copy operations.
+///
+/// This method is using virtual dispatch as a trick to obtain T, such that
+/// transform can be called with any T.
+template <class T>
+void DataModel<T>::copy(const VariableConstView &src,
```

Comment:
As future clean up, should we look at separating off `DataModel` so it can be unit tested in isolation?

Reply:
Note that it is likely the code here and in relation to this will be refactored more in the next steps of #1281.